design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,475 @@
+kind: ControllerRevision
+metadata:
+  labels:
+    controller-revision-hash: a75588698
+  name: frontend-a75588698
+  namespace: default
+  ownerReferences:
+  - apiVersion: core.oam.dev/v1alpha2
+    blockOwnerDeletion: true
+    controller: true
+    kind: Component
+    name: frontend
+revision: 2
+data:
+  spec:
+    workload:
+      apiVersion: core.oam.dev/v1alpha2
+      kind: ContainerizedWorkload
+      spec:
+        containers:
+        - name: my-cool-workload
+          image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+          cmd:
+          - "bash top"
+```
+
+Now that the basic controller revision are in place, we can use them in Trait and ApplicationConfiguration.
+
+### A real running workload instance will have the same name with revision
+
+If we create an `ApplicationConfiguration` with definition : 
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: frontend
+```
+
+The real workload instance will be created with the same name as the controllerRevision.
+
+```yaml
+  apiVersion: core.oam.dev/v1alpha2
+  kind: ContainerizedWorkload
+  metadata:
+    name: frontend-a75588698
+  spec:
+    containers:
+    - name: my-cool-workload
+      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+      cmd:
+      - "bash lscpu"
+```
+
+
+### ControllerRevision could be referenced by Trait 
+
+Some traits can control version of component which means they can specify a component revision and won't be affected
+by the change of component.
+
+For these traits, they have to declare `revision` field in TraitDefinition like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: fancytraits.core.oam.dev
+spec:
+  revision:
+    alwaysControl: true
+    revisonPath:
+      - "spec.revision"
+  definitionRef:
+    name: fancytraits.core.oam.dev
+```
+
+#### Revision field rule
+
+Versioning mechanism will be related with `revision` field.
+
+- If no traits binding with a component have revision field, OAM runtime will always update workload directly.
+- If `alwaysControl` field is true, OAM runtime will always create a new workload when component revision changed
+  and leave the trait to control two workloads rolling update.
+- If `alwaysControl` field is false, this means the trait will control only if revision path specified. OAM runtime
+  will check if no `revisionPath` specified, then it will update workload directly, otherwise it will create new workload. 
+ 
+
+So when we use the trait like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: frontend
+      traits:
+        - trait:
+            apiVersion: core.oam.dev/v1alpha2
+            kind: FancyTrait
+            spec:
+              revision: frontend-c8bb659c5 
```

Comment:
is the `alwaysControl` field optional?

Reply:
Yes, it's optional, but you are commenting the wrong place, right? Only traitDefinition has alwaysControl field.